### ISSUE TITLE
Queue take race condition

### ISSUE THREAD

**Issue by RaasAhsan** [scalaz/scalaz-zio#447]
While investigating #438, I discovered a bug that I was able to narrow down to a race condition present in the current implementation of `Queue`. The consequence of the race condition is that elements may be removed from the queue out-of-order in which they were enqueued.

The race condition manifests here (https://github.com/scalaz/scalaz-zio/blob/master/core/shared/src/main/scala/scalaz/zio/Queue.scala#L166 between lines 166 and 167). When a taker thread reaches line 166, it offers a `Promise` callback for the taker to the `takers` queue. On line 167, `unsafeCompleteTakers` is invoked to remove any elements in `queue` that were added after the `poll` on line 155. 

Imagine a taker thread T is at line 166 in its execution. From the perspective of T, another thread S concurrently enqueues an element A to `queue`. T adds its `Promise` to the `takers` queue on line 166.  Concurrently, right before T begins line 167 line 167, S offers another element B to the queue. The `offer` invocation finds the `Promise` that T enqueued to the `takers` queue, removes it and completes it with element B. Thread T now invokes `unsafeCompleteTakers` and observes that element A is available in `queue`, but the taker it enqueued to `takers` a line before has already been consumed. The offering of element A occurs causally before the offering of element B, but B is consumed first.

Here is a code example that illustrates this race condition:

```scala
for {
  queue <- Queue.bounded[Int](1)
  put    =  queue.offer(3).flatMap(_ => queue.offer(4))
  take   = queue.take.flatMap(a => queue.take.map(b => List(a, b)))
  _     <- put.fork
  list  <- take
} yield list
```

Over many executions, you may observe that `list` equals `List(4, 3)` instead of `List(3, 4)`. I was looking for a simple fix to eliminate the race condition, but I didn't come across an easy solution. I think the invariant that needs to be satisfied is that either `queue` or `takers` can be non-empty, but not both.

This race condition also manifests for `offer` (https://github.com/scalaz/scalaz-zio/blob/master/core/shared/src/main/scala/scalaz/zio/Queue.scala#L96) but if `Queue` is not guaranteed to be fair to consumers, then it isn't a problem.

Let me know if you need more information or if I can help out with a fix in any way. :) @jdegoes I'm not sure who to ping for this, could you make sure the right people see it?

**Comment by ghostdogpr** [comment]
Hey, thanks for the detailed report. I'm the one who did this in the first place 😕
The queue is intended to be fair, even though it's getting tricky now that we use several mutable queues (the previous implementation was wrapping everything in Ref.modify so there was no concurrency, which means it was also much slower).

One thing we can do in `offer` is try to poll takers only if the queue is empty.
```
val pTakers = if (queue.isEmpty()) unsafePollN(takers, as.size) else List.empty
```
That way, if the queue is not empty, it will force enqueuing items so that right after `unsafeCompleteTakers` processes them in the right order.

I ran your code 10 million times and the issue didn't appear.

**Comment by RaasAhsan** [comment]
I tried that out too, didn't see any issues! 😄  Though, just to confirm, the race condition is still there, right?

As far as fairness is concerned, it makes sense that techniques that improve performance would degrade fairness. I can identify several sources of unfairness in the current implementation, and "fixing" them would definitely degrade performance. If a fair queue is desired, perhaps that could be exposed as another type or strategy?

**Comment by ghostdogpr** [comment]
Yeah I think there's still a very minimal chance that it happens.

I guess it's a compromise to find between fairness and performance. Another approach is Monix which is optimized for speed but with low fairness (no takers and putters queue, every taker and putter keep being retried without order until they can be processed).

I would say that the current implementation maintains both high performance and high fairness (but not absolute). If we find fixes like this one we should totally apply them, but if the performance penalty is too important it may be not worth it.

@jdegoes what's your take on this?

**Comment by jdegoes** [comment]
@ghostdogpr I agree. "High performance" and "high fairness" is the sweet spot for 90% of applications.

**Comment by ghostdogpr** [comment]
I included that fix in #443

**Comment by ghostdogpr** [comment]
Closed by #443
